src/ChannelServer/World/Entities/Components/Buffs.cs

```diff
@@ -0,0 +1,239 @@
+		/// <returns></returns>
+		public bool Remove(BuffId buffId)
+		{
+
+			lock (_buffs) {
+				var buff = _buffs[buffId];
+				if (_buffs.Remove(buffId))
+					// Update client...
+					Send.ZC_BUFF_REMOVE(this.Character, buff);
+			}
+
+			return true;
+		}
+
+		/// <summary>
+		/// Removes buff with given id, returns false if it
+		/// didn't exist. Updates the client on success.
+		/// </summary>
+		/// <param name="buff"></param>
+		/// <returns></returns>
+		public bool Remove(Buff buff)
+		{
+			Log.Debug("Removing buff {0} which expired at {1}.", buff.Data.ClassName, buff.RemovalTime);
+			lock (_buffs)
+			{
+				if (!_buffs.Remove(buff.Id))
+					return false;
+				// Update client...
+				Send.ZC_BUFF_REMOVE(this.Character, buff);
+			}
+
+			return true;
+		}
+
+		/// <summary>
+		/// Add a Buff or Update a Buff if it exists
+		/// with a given id.
+		/// </summary>
+		/// <param name="buffId"></param>
+		public void AddOrUpdate(BuffId buffId)
+		{
+			if (!_buffs.ContainsKey(buffId)) {
+				var buff = new Buff(this.Character, buffId);
+				Add(buff);
+			} else
+			{
+				Update(buffId);
+			}
+		}
+
+		/// <summary>
+		/// 
+		/// </summary>
+		/// <param name="buffId"></param>
+		public void Update(BuffId buffId)
+		{
+			lock (_buffs)
+			{
+				var buff = _buffs[buffId];
+				buff.IncreaseOverbuff();
+				buff.RemovalTime = DateTime.Now.AddMilliseconds(buff.Data.Duration);
+				Send.ZC_BUFF_UPDATE(this.Character, buff);
+			}
+		}
+
+		/// <summary>
+		/// Returns buff with given id, or null if it didn't
+		/// exist.
+		/// </summary>
+		/// <param name="buffId"></param>
+		/// <returns></returns>
+		public Buff Get(BuffId buffId)
+		{
+			lock (_buffs)
+			{
+				_buffs.TryGetValue(buffId, out var result);
+				return result;
+			}
+		}
+
+		/// <summary>
+		/// Returns buff with given id via out, returns false if the
+		/// buff wasn't found.
+		/// </summary>
+		/// <param name="buffId"></param>
+		/// <param name="buff"></param>
+		/// <returns></returns>
+		public bool TryGet(BuffId buffId, out Buff buff)
+		{
+			buff = this.Get(buffId);
+			return buff != null;
+		}
+
+		/// <summary>
+		/// Returns buff with given class name, or null if it didn't
+		/// exist.
+		/// </summary>
+		/// <param name="buffId"></param>
+		/// <returns></returns>
+		public Buff Get(string buffClassName)
+		{
+			lock (_buffs)
+				return _buffs.Values.FirstOrDefault(a => a.Data.ClassName == buffClassName);
+		}
+
+		/// <summary>
+		/// Returns a list with all buffs.
+		/// </summary>
+		/// <returns></returns>
+		public Buff[] GetList()
+		{
+			lock (_buffs)
+				return _buffs.Values.ToArray();
+		}
+
+		/// <summary>
+		/// Returns true if the buff exists.
+		/// </summary>
+		/// <param name="buffId"></param>
+		/// <returns></returns>
+		public bool Has(BuffId buffId)
+		{
+			lock (_buffs)
+				return _buffs.ContainsKey(buffId);
+		}
+
+		/// <summary>
+		/// Returns current level of given buff, returns 0 if buff
```

Comment:
You say "current level", but it always returns the level from the data, can the level actually change?

Reply:
Instead of level changing, I think over buff counter is the right way to go about it. Because buffs have different effects on stacks, maybe rename Over Buff Counter to Buff Stacks? So maybe level is useless in buffs.

Reply:
Do I have that right, that buffs can stack, and that that changes their effect?

Reply:
I'm basing it off buffs I've seen, thus far. 
https://treeofsavior.com/page/class/view.php?c=Scout
Oblique Fire - 5 stacks maximum: increases your movement speed when 3 stacks or lower and decreases the enemy's movement speed when 4 stacks or more.

Reply:
Hm. Yea, that does sound like that might be how it works.